File: apps/wazuh/wazuhApiAuthenticator.py
#!/usr/bin/env python3
import json
import requests
import urllib3
from base64 import b64encode

# Disable insecure HTTPS warnings (for self-signed SSL certificates)
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

CONFIG_FILE = 'config.json'

def load_config():
    """Load configuration settings from config.json."""
    try:
        with open(CONFIG_FILE, 'r') as f:
            return json.load(f)
    except Exception as e:
        print(f"Error loading configuration: {e}")
        exit(1)

def save_config(config):
    """Write updated configuration settings back to config.json."""
    try:
        with open(CONFIG_FILE, 'w') as f:
            json.dump(config, f, indent=4)
    except Exception as e:
        print(f"Error saving configuration: {e}")
        exit(1)

def confirm_config(config):
    """
    Confirm configuration with the user, allowing them to update values if necessary.
    Keys confirmed: protocol, host, port, user, password, and login_endpoint.
    """
    keys = ['protocol', 'host', 'port', 'user', 'password', 'login_endpoint']
    print("Current configuration:")
    for key in keys:
        print(f"  {key}: {config.get(key)}")
    answer = input("Are these values correct? (y/n): ").strip().lower()
    if answer != 'y':
        print("Enter new values (press Enter to keep the current value):")
        for key in keys:
            current_value = config.get(key)
            new_value = input(f"  {key} [{current_value}]: ").strip()
            if new_value:
                config[key] = new_value
        save_config(config)
        print("Configuration updated.\n")
    return config

def authenticate(config):
    """
    Authenticate to the Wazuh API using Basic Authentication.
    Returns the JWT token retrieved from the API.
    """
    protocol      = config.get('protocol', 'https')
    host          = config.get('host', 'localhost')
    port          = config.get('port', '55000')
    user          = config.get('user')
    password      = config.get('password')
    login_endpoint= config.get('login_endpoint', 'security/user/authenticate')
    
    login_url = f"{protocol}://{host}:{port}/{login_endpoint}"
    basic_auth = f"{user}:{password}".encode()
    login_headers = {
        'Content-Type': 'application/json',
        'Authorization': f'Basic {b64encode(basic_auth).decode()}'
    }

    print("\nLogin request ...\n")
    try:
        response = requests.post(login_url, headers=login_headers, verify=False)
    except requests.exceptions.RequestException as e:
        print(f"Login request failed: {e}")
        exit(1)

    if response.status_code != 200:
        print(f"Login failed ({response.status_code}): {response.text}")
        exit(1)

    try:
        token = response.json()['data']['token']
    except Exception as e:
        print(f"Error parsing login response: {e}")
        exit(1)

    print("Token received:\n", token)
    return token

def get_api_info(config, token):
    """
    Retrieve API information from the Wazuh API.
    """
    protocol = config.get('protocol', 'https')
    host     = config.get('host', 'localhost')
    port     = config.get('port', '55000')
    url      = f"{protocol}://{host}:{port}/?pretty=true"
    headers  = {
        'Content-Type': 'application/json',
        'Authorization': f'Bearer {token}'
    }
    
    print("\nGetting API information:\n")
    try:
        response = requests.get(url, headers=headers, verify=False)
        print(response.text)
    except requests.exceptions.RequestException as e:
        print(f"Error retrieving API information: {e}")

def get_agents_status(config, token):
    """
    Retrieve the agents status summary from the Wazuh API.
    """
    protocol = config.get('protocol', 'https')
    host     = config.get('host', 'localhost')
    port     = config.get('port', '55000')
    url      = f"{protocol}://{host}:{port}/agents/summary/status?pretty=true"
    headers  = {
        'Content-Type': 'application/json',
        'Authorization': f'Bearer {token}'
    }
    
    print("\nGetting agents status summary:\n")
    try:
        response = requests.get(url, headers=headers, verify=False)
        print(response.text)
    except requests.exceptions.RequestException as e:
        print(f"Error retrieving agents status summary: {e}")

def main():
    # Load and confirm configuration
    config = load_config()
    config = confirm_config(config)

    # Authenticate and get a JWT token
    token = authenticate(config)

    # Save the token into configuration for future reference
    config['jwt_token'] = token
    save_config(config)

    # Perform API calls using the token
    get_api_info(config, token)
    get_agents_status(config, token)

    print("\nEnd of the script.\n")

if __name__ == "__main__":
    main()

File: apps/wazuh/wazuhApiAuthenticator.go
package main

import (
	"bufio"
	"bytes"
	"crypto/tls"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"strings"
)

// Config represents the configuration file structure.
type Config struct {
	Protocol      string `json:"protocol"`
	Host          string `json:"host"`
	Port          string `json:"port"`
	User          string `json:"user"`
	Password      string `json:"password"`
	LoginEndpoint string `json:"login_endpoint"`
	JWTToken      string `json:"jwt_token,omitempty"`
}

const configFile = "config.json"

// loadConfig loads configuration settings from config.json.
func loadConfig() Config {
	var cfg Config
	data, err := ioutil.ReadFile(configFile)
	if err != nil {
		fmt.Printf("Error loading configuration: %v\n", err)
		os.Exit(1)
	}
	if err := json.Unmarshal(data, &cfg); err != nil {
		fmt.Printf("Error parsing configuration: %v\n", err)
		os.Exit(1)
	}
	return cfg
}

// saveConfig writes updated configuration settings back to config.json.
func saveConfig(cfg Config) {
	data, err := json.MarshalIndent(cfg, "", "    ")
	if err != nil {
		fmt.Printf("Error encoding configuration: %v\n", err)
		os.Exit(1)
	}
	if err := ioutil.WriteFile(configFile, data, 0644); err != nil {
		fmt.Printf("Error saving configuration: %v\n", err)
		os.Exit(1)
	}
}

// promptInput prompts the user for a value and returns the new value or the current one if nothing is entered.
func promptInput(prompt, current string) string {
	reader := bufio.NewReader(os.Stdin)
	if current != "" {
		fmt.Printf("  %s [%s]: ", prompt, current)
	} else {
		fmt.Printf("  %s: ", prompt)
	}
	input, _ := reader.ReadString('\n')
	input = strings.TrimSpace(input)
	if input == "" {
		return current
	}
	return input
}

// confirmConfig displays the current configuration and allows the user to update values.
func confirmConfig(cfg Config) Config {
	keys := []string{"protocol", "host", "port", "user", "password", "login_endpoint"}
	fmt.Println("Current configuration:")
	fmt.Printf("  protocol:      %s\n", cfg.Protocol)
	fmt.Printf("  host:          %s\n", cfg.Host)
	fmt.Printf("  port:          %s\n", cfg.Port)
	fmt.Printf("  user:          %s\n", cfg.User)
	fmt.Printf("  password:      %s\n", cfg.Password)
	fmt.Printf("  login_endpoint:%s\n", cfg.LoginEndpoint)

	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Are these values correct? (y/n): ")
	answer, _ := reader.ReadString('\n')
	answer = strings.TrimSpace(strings.ToLower(answer))

	if answer != "y" {
		fmt.Println("Enter new values (press Enter to keep the current value):")
		cfg.Protocol = promptInput("protocol", cfg.Protocol)
		cfg.Host = promptInput("host", cfg.Host)
		cfg.Port = promptInput("port", cfg.Port)
		cfg.User = promptInput("user", cfg.User)
		cfg.Password = promptInput("password", cfg.Password)
		cfg.LoginEndpoint = promptInput("login_endpoint", cfg.LoginEndpoint)
		saveConfig(cfg)
		fmt.Println("Configuration updated.\n")
	}
	return cfg
}

// AuthResponse models the JSON response from the login endpoint.
type AuthResponse struct {
	Data struct {
		Token string `json:"token"`
	} `json:"data"`
}

// authenticate performs a login request using Basic Authentication to retrieve the JWT token.
func authenticate(cfg Config) string {
	// Build the URL.
	url := fmt.Sprintf("%s://%s:%s/%s", cfg.Protocol, cfg.Host, cfg.Port, cfg.LoginEndpoint)

	// Create the basic auth header.
	authStr := fmt.Sprintf("%s:%s", cfg.User, cfg.Password)
	encodedAuth := base64.StdEncoding.EncodeToString([]byte(authStr))
	loginHeaders := map[string]string{
		"Content-Type":  "application/json",
		"Authorization": fmt.Sprintf("Basic %s", encodedAuth),
	}

	fmt.Println("\nLogin request ...\n")
	// Create a custom HTTP client that skips certificate verification.
	tr := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}
	client := &http.Client{Transport: tr}

	req, err := http.NewRequest("POST", url, nil)
	if err != nil {
		fmt.Printf("Error creating login request: %v\n", err)
		os.Exit(1)
	}
	for k, v := range loginHeaders {
		req.Header.Set(k, v)
	}

	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("Login request failed: %v\n", err)
		os.Exit(1)
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		fmt.Printf("Error reading login response: %v\n", err)
		os.Exit(1)
	}

	if resp.StatusCode != http.StatusOK {
		fmt.Printf("Login failed (%d): %s\n", resp.StatusCode, string(body))
		os.Exit(1)
	}

	var authResp AuthResponse
	if err := json.Unmarshal(body, &authResp); err != nil {
		fmt.Printf("Error parsing login response: %v\n", err)
		os.Exit(1)
	}

	token := authResp.Data.Token
	if token == "" {
		fmt.Println("Error: token not found in login response")
		os.Exit(1)
	}

	fmt.Println("Token received:\n", token)
	return token
}

// getAPIInfo retrieves API information from the Wazuh API.
func getAPIInfo(cfg Config, token string) {
	url := fmt.Sprintf("%s://%s:%s/?pretty=true", cfg.Protocol, cfg.Host, cfg.Port)
	headers := map[string]string{
		"Content-Type":  "application/json",
		"Authorization": fmt.Sprintf("Bearer %s", token),
	}
	fmt.Println("\nGetting API information:\n")
	makeGetRequest(url, headers)
}

// getAgentsStatus retrieves the agents status summary from the Wazuh API.
func getAgentsStatus(cfg Config, token string) {
	url := fmt.Sprintf("%s://%s:%s/agents/summary/status?pretty=true", cfg.Protocol, cfg.Host, cfg.Port)
	headers := map[string]string{
		"Content-Type":  "application/json",
		"Authorization": fmt.Sprintf("Bearer %s", token),
	}
	fmt.Println("\nGetting agents status summary:\n")
	makeGetRequest(url, headers)
}

// makeGetRequest is a helper to perform GET requests and print the response.
func makeGetRequest(url string, headers map[string]string) {
	tr := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}
	client := &http.Client{Transport: tr}
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		fmt.Printf("Error creating GET request: %v\n", err)
		return
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("Error retrieving data: %v\n", err)
		return
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Printf("Error reading response: %v\n", err)
		return
	}
	fmt.Println(string(body))
}

func main() {
	// Load and confirm configuration.
	cfg := loadConfig()
	cfg = confirmConfig(cfg)

	// Authenticate and get a JWT token.
	token := authenticate(cfg)
	cfg.JWTToken = token
	saveConfig(cfg)

	// Perform API calls using the token.
	getAPIInfo(cfg, token)
	getAgentsStatus(cfg, token)

	fmt.Println("\nEnd of the script.\n")
}
